Photo Proof App

ilynne opened the pull request:
I'm going to immediately close this pull request. It's a draft.

/jobs/:id/photo_proofs/new:

No authentication is required to access this page. The file tag uses the capture attribute, should capture from the device's camera.
![kotis_print_shop](https://cloud.githubusercontent.com/assets/3128785/9370594/59d7ccee-4686-11e5-8428-0de03308a337.png)

After proof is uploaded, confirmation is displayed. The image only shows when a flash message is present.
![kotis_print_shop](https://cloud.githubusercontent.com/assets/3128785/9370711/26ba5218-4687-11e5-997a-295f7f34adc3.png)

This bare-bones email is sent.
![photo_proof_for_job_115413](https://cloud.githubusercontent.com/assets/3128785/9370875/9848f24e-4688-11e5-9736-be0b901de4f2.png)

The approve/reject link requires an authenticated user.
![kotis_print_shop](https://cloud.githubusercontent.com/assets/3128785/9370903/e84268ca-4688-11e5-8412-59d0402b846c.png)

There is a photo proofs page, which also requires an authenticated user.
![kotis_print_shop](https://cloud.githubusercontent.com/assets/3128785/9370928/128accda-4689-11e5-85cc-93f513e4fa8c.png)

Individual proofs can also be viewed by authenticated users. The "Resend Email" button would resend the proof. Not sure we need it, not sure we don't.
![kotis_print_shop](https://cloud.githubusercontent.com/assets/3128785/9370945/370058dc-4689-11e5-8964-aedf15d9c13a.png)

closes #1

becker1 commented:
Why are we creating a PR if you are going to close it immediately?

ilynne commented:
I did it as a draft.

becker1 commented:
I know, I'm wondering why we need that? Anyone can do a comparison of branches. If you want people to look at it, you should leave it open. If you don't want anyone to look at it, why are you creating it?

ilynne commented:
I was almost done with it when I got your email. I did not want to abandon all the writing. I need to change the approval.

becker1 commented:
Just don't do too much code without people having a chance to review it, we should do it pretty early in the process in case anything major needs to get changed

becker1 commented:
(Meaning, I sent you the email assuming we would do it after this branch is merged; we don't need to keep adding more and more to this one)

ilynne commented:
OK, reopening, then. The Approve button/dropdown can be a new issue.

jonfriese commented on app/controllers/admin_controller.rb:
```diff
@@ -2,14 +2,22 @@ class AdminController < ApplicationController
-
   private
 
   def authorize_employees!
```
Should this be singular `authorize_employee!`?

jonfriese commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,95 @@
+  before_action :set_photo_proof, except: [:create, :new, :index]
+
+  def index
+    @photo_proofs = PhotoProof.includes(:job).where(job_id: @job.id)
```
Could you do `@job.photo_proofs` instead?

jonfriese commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,95 @@
+  end
+
+  def new
+    set_photo_proof if params[:id] && flash.present?
```
It seems like `params[:id]` would never be nil so that check may be unnecessary.

ilynne replied:
It is nil if someone clicked the "Upload" button without submitting a file first.

jayfredlund replied:
Why would this be called in the `new` action?  `new` should be reserved for new objects -- `show` should be used if there's already an object.

Then, in a before_action on `:new`, you'd do `@photo_proof = PhotoProof.new`, use this object in the file uploader form, and then you can use this object for the object creation when posting the form to `:create`.

jonfriese commented on lib/s3_file_uploader.rb:
```diff
@@ -0,0 +1,38 @@
+  def upload
+    @uploader.upload_file(@file, acl: 'public-read')
+  rescue Aws::S3::Errors::ServiceError => e
+    log_and_notify_airbrake(e)
```
I think you may want to return false here. So that the photo_proofs_controller#create flashes an error message when an error is rescued.

jonfriese commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,95 @@
+
+      redirect_to new_job_photo_proof_path(job_id: @job.id, id: @photo_proof.id), notice: "Photo Proof sent:"
+    else
+      flash.now[:notice] = 'There was an error'
```
Not sure what this would be, but maybe add something a little more descriptive. Or maybe what the user should do when this fails.

jonfriese commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,95 @@
+
+  def resend
+    notice = "Something went wrong." unless photo_proof_email_resend
+    notice = "Photo proof resent."
```
Won't these two lines always result in notice being `Photo proof resent.`

jonfriese commented on app/helpers/jobs/photo_proofs_helper.rb:
```diff
@@ -0,0 +1,18 @@
+                             "-#{photo_proof.approved.to_s}"
+    when false
+      icon, message, style = "close",
+                             "Rejected by #{User.find_by(kotisauth_id: photo_proof.reviewed_by_id).full_name}",
```
You could set a variable `user_full_name` before the case statement to clean this up a little.

ilynne replied:
I'm going to leave this as-is because "Approved" is going to change to a status (next issue) and this helper will change.

jonfriese replied:
:+1:

bonhamjp commented on app/views/layouts/mailer.html.erb:
```diff
@@ -0,0 +1,9 @@
+  <head>
+  </head>
+  <body>
+    <div style='font-family: "Open Sans", sans-serif;'>
```
Why use an inline style here?

ilynne replied:
It's a mail template. Stylesheets don't always work.

bonhamjp commented on app/views/photo_proof_mailer/notify_contacts_of_photo_proof.html.erb:
```diff
@@ -0,0 +1,5 @@
+<div id="photo-proof-container" style="max-width: 100%">
+  <%= image_tag(@photo_proof.public_url, style: "max-width: 100%;") %>
+</div>
+<p><%= link_to "Approve or reject this photo proof", approve_job_photo_proof_url(id: @photo_proof.id, job_id: @photo_proof.job_id), method: :get %></p>
```
This line breaks the 100 character max

bonhamjp commented on config/routes.rb:
```diff
@@ -1,9 +1,32 @@
+
+  resources :jobs, only: :show do
+    resources :photo_proofs,
+      only: [:index, :show, :new, :create, :update],
```
I than you can do
```Ruby
  except: [:edit, :destroy]
```

ilynne replied:
Yep. I started with only :show, but then more resources were added.

jayfredlund commented:
Instead of using S3 directly, can we use Carrierwave for the uploads so we're consistent with all of the other apps?

jonfriese commented on app/models/photo_proof.rb:
```diff
@@ -0,0 +1,12 @@
+  default_scope { order(created_at: :desc) }
+
+  def approvable?
+    approved.nil?
```
I assume this will never be false, but if you wanted to handle all falsey values you could change this to `!approved`.

ilynne replied:
This is going to change, too. "Approved" will probably go away entirely, since there will be a string status. In this soon-to-be-exterminated case, approved could be true, false, or nil. The record is approvable if no value has been set yet.

jonfriese replied:
Oh yeah, duh.

jonfriese commented on lib/s3_file_uploader.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  def upload
+    @uploader.upload_file(@file, acl: 'public-read')
```
You can just do `uploader` here since you have an attr_reader method for `@uploader`. Same with all the other ones.

jayfredlund commented on lib/s3_file_uploader.rb:
```diff
@@ -0,0 +1,39 @@
+
+class S3FileUploader
```
I'm not going to comment on this PR much yet, but with Carrierwave installed (and configured to use fog), you should be able to get rid of this whole file, remove the `Upload` model, and get rid of a bunch of logic in the photo_proofs controller.

alkx commented on lib/s3_file_uploader.rb:
```diff
@@ -0,0 +1,39 @@
+    "#{DateTime.now.strftime("%s")}-#{sanitized_filename}"
+  end
+
+  def log_and_notify_airbrake(e)
```
Is this something we should include in a module? The same method is in `job_details.rb`

jayfredlund replied:
This entire file is going to go away, FYI, once carrierwave is installed and set up.

Separately, we don't need to document anything to the `Rails.logger` -- notifying airbrake is what we care about.

ilynne commented:
OK, comments addressed, lots of changes. New API in use, Carrierwave handling the file uploads.

This should not be put into actual production until issues #3 and #4 are merged in.

becker1 commented:
Do we have any way of leaving these changes out and putting them in a different branch? After people do code review we should not be adding more changes to the branch, otherwise it is too difficult to review.

ilynne commented:
Do you mean issues #3 and #4? I was not going to add them to this branch, I was going to do separate branches for them.

becker1 commented:
Oh yeah, I misread your comment. That should be fine then. I just want to make sure we don't add a ton of new features to this branch.

jayfredlund commented:
FYI, I'm going to take a crack at this PR in a little bit.

jayfredlund commented on app/assets/javascripts/job.js.coffee:
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Let's delete any of these files that we aren't using.  We'll add these in as we add features to the app.

jayfredlund commented on app/assets/stylesheets/jobs/photo_proofs.css.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the jobs/photo_proofs controller here.
```
Same with unused css files -- remove them from here and the main import directives.

When we add styles to these pages, we'll add the files in at the same time.

becker1 replied:
I'm just scanning the comments that are still left - doesn't look like this got addressed?

jayfredlund commented on app/assets/stylesheets/styles/photo_proofs.css.scss:
```diff
@@ -0,0 +1,60 @@
+    vertical-align: top;
+  }
+}
+.status-approved {
```
These styles would be better implemented as nested with subclasses:
```SCSS
.class-of-object-type { /* the naming of this class should be something different.  What is the type of the element?  "status-message" maybe? */
  color: $kotis-gray;
  .approved {
    color: $kotis-green;
  }
  .rejected {
    color: $kotis-red;
  }
}
```
I'll comment on the code where this should be changed as well.

ilynne replied:
This I will leave as-is because in issue #3 the helper, views and styles will change. But I will refer back to this suggestion.

becker1 replied:
Also this one and one more. @ilynne can you just review the remaining comments in github and make sure there isn't anything you still need to address? It's possible the line of code that was commented on simply wasn't changed in some of the cases.

jayfredlund commented on app/assets/stylesheets/styles/photo_proofs.css.scss:
```diff
@@ -0,0 +1,60 @@
+    margin: 1em 0 0;
+  }
+  img {
+    max-width: 100%;
```
Do you need to specify these?  You can use the class `pure-img` on the actual image to make them responsive.

jayfredlund commented on app/controllers/admin_controller.rb:
```diff
@@ -2,14 +2,20 @@ class AdminController < ApplicationController
   include KotisauthGem::HelperMethods
   include RoleChecker
 
-  before_action :authenticate_user!
```
Presumably, anything in the admin controller should run these, right?  I think it makes sense to keep this in the `AdminController` rather than every controller that inherits from it..

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+class Jobs::PhotoProofsController < AdminController
+  before_action :authenticate_user!, except: [:new, :create]
```
Why would we want to allow unauthenticated users to create photo proofs?

ilynne replied:
The users might be temporary workers who are not signed in. The photo proofs would be sent from a device that is shared. We do need some sort of check to prevent anyone in the world from sending a photo of their cat as a photo proof.

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+  before_action :authorize_employee!, except: [:new, :create]
+
+  before_action :set_job
+  before_action :redirect_if_missing_file, only: :create
```
Instead of doing this in a before action, can't we just try to create a photo proof with no upload and if it can't be created, render the form with an error that says "no file?".  This seems like we're doing a lot of manual checking that Carrierwave and model validations really help with.

For example, we don't need to do the "redirect if missing file" if we require the presence of an image on the PhotoProof model to save.  

Then we can just call 
```Ruby
if @photo_proof.save
  # do something successful
else
  render :new
end
```
 
Curious, have you watched [the carrierwave railscast](http://railscasts.com/episodes/253-carrierwave-file-uploads?view=asciicast)?  It's really helpful and will still likely simplify a whole lot of the file handling stuff you're doing between the `new` and `create` actions?

ilynne replied:
Yes, a long time ago -- it's over four years old. But I watched it again anyway. Yes, using carrierwave made a lot of this stuff go away.

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+  end
+
+  def create
+    @photo_proof = PhotoProof.new(job: @job, photograph: params[:file].tempfile)
```
This should be handled using whitelisted params and the object should be used in the form.
```Ruby
def create 
  if @photo_proof.create(photo_proof_params)
    # do successful stuff, like send email
  else
    render :new
  end
end

private

def photo_proof_params
  params.require(:photo_proof).permit(:photograph)
end
```

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+    @photo_proof = PhotoProof.new(job: @job, photograph: params[:file].tempfile)
+    if @photo_proof.save && photo_proof_email_send
+      flash[:notice] = "Photo Proof sent:"
+      redirect_to new_job_photo_proof_path(job_id: @job.id, id: @photo_proof.id)
```
Instead of redirecting to new, this should be the "show" action.

ilynne replied:
I was going to new because I was assuming that the unknown user might need feedback that the proof was successful and might want to send another... but then again, I don't know how the unknown user gets to the new page anyway, so I'll change this.

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+
+  def update
+    if @photo_proof.approvable?
+      params[:photo_proof][:approved] = true if params[:commit].downcase == "approve"
```
We should be using the object we're updating in the form -- so if you use the `@photo_proof` object in the form, you can whitelist the params and just update these accordingly (with an approved radio/checkbox button) and (checked/validated here).

ilynne replied:
This is actually going to change in issue #3, so I will leave it as is for now and keep these comments in mind for that issue.

jayfredlund replied:
I just want to make sure that the form gets changed from using `form_tag` to `form_for` with the PhotoProof object in this PR.

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+      params[:photo_proof][:approved] = true if params[:commit].downcase == "approve"
+      params[:photo_proof][:approved] = false if params[:commit].downcase == "reject"
+      params[:photo_proof][:reviewed_by_id] = @user_session.kotisauth_id
+      @photo_proof.update!(photo_proof_params)
```
Do we want to throw an exception here?  

We should probably just use `.update` instead of `.update!` since we're not rescuing from exceptions.

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+    else
+      flash[:error] = "This Photo Proof has already been reviewed."
+    end
+      redirect_to action: :show
```
This is indented one too many :)

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+    @job = Job.find_or_create_by(id: params[:job_id])
+  end
+
+  def set_photo_proof
```
This should be used on the `:show` action and probably called something like `get_photo_proof`.  And if it can't be found, should redirect the user somewhere.

For the `:new` action, you should use something like:
```Ruby
def set_photo_proof
  @photo_proof = PhotoProof.new(job: @job)
end
```

jayfredlund commented on app/controllers/jobs/photo_proofs_controller.rb:
```diff
@@ -0,0 +1,81 @@
+  end
+
+  def photo_proof_params
+    params.require(:photo_proof).permit(:photo_proof, :approved, :reviewed_by_id, :comments)
```
Why are we nesting `:photo_proof` inside of a `:photo_proof` object?

If you add the `:photograph` uploader here, a lot of the stuff related to the params file handling can go away (if we're using carrierwave).

jayfredlund commented on app/controllers/jobs_controller.rb:
```diff
@@ -0,0 +1,18 @@
+  private
+
+  def set_job
+    @job = Job.where(id: params[:id]).first_or_create!
```
How about we use a find_or_create?  The job id should be unique.

jayfredlund commented on app/controllers/jobs_controller.rb:
```diff
@@ -0,0 +1,18 @@
+class JobsController < AdminController
+  before_action :authenticate_user!
```
These should all be in the AdminController.  For actions where we want to skip the authentication, you can use `skip_before_action` (like you suggested of GH :+1: )

jayfredlund commented on app/helpers/application_helper.rb:
```diff
@@ -11,4 +11,8 @@ def flash_class(level)
   end
 
+  def local_datetime(date)
+    date.in_time_zone('Pacific Time (US & Canada)').strftime("%-d/%-m/%y %l:%M%P")
```
We should set the Application wide time zone so we don't have to use this format.

In application.rb:
```Ruby
config.time_zone = 'Pacific Time (US & Canada)'
```

Then you can just make this helper method format the time.

jayfredlund commented on app/helpers/job_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module JobHelper
```
We can remove this file since it's not used.

jayfredlund commented on app/helpers/jobs/photo_proofs_helper.rb:
```diff
@@ -0,0 +1,18 @@
+module Jobs::PhotoProofsHelper
```
We should use the newer/preferred syntax for this:
```Ruby
module Jobs
  module PhotoProofsHelper
```

jayfredlund commented on app/helpers/jobs/photo_proofs_helper.rb:
```diff
@@ -0,0 +1,18 @@
+    else
+      icon, message, style = "minus", "Awaiting Approval", nil
+    end
+    "<span class=\"status-approved#{style}\"><i class=\"fa fa-#{icon}\"></i> #{message}</span>"
```
Let's use the `content_tag` helper here:
```Ruby
content_tag :span, class: "approval-message #{style}" do
  #use the icon helper to generate the icon
  icon("#{icon}") + message
end
```

This is where your "style" should now specify the class to apply (related to the nested class definitions i commented on above).

ilynne replied:
Yep. And I'll leave as-is for this PR and apply this comment if necessary to issue #3.

jayfredlund commented on app/mailers/application_mailer.rb:
```diff
@@ -0,0 +1,4 @@
+class ApplicationMailer < ActionMailer::Base
+  default from: "from@example.com"
```
We should probably change the default to "noreply@kotsidesign.com"

jayfredlund commented on app/uploaders/photo_proof_uploader.rb:
```diff
@@ -0,0 +1,25 @@
+  protected
+
+  def set_filename
+    model.filename = "#{SecureRandom.uuid}.#{file.extension}"
```
This is probably fine for the name for now since it's internal -- but we may eventually want to change this so the name is something like "photo-proof-#{version_number}".  Where version_number would be a count of the number of photo proofs associated with this job.

Nothing to do for now.

jayfredlund commented on app/views/jobs/photo_proofs/_upload_form.html.erb:
```diff
@@ -0,0 +1,6 @@
+<%= form_tag job_photo_proofs_path(job_id: @job.id), id: "photo-proof-upload-form", enctype: 'multipart/form-data', class: "pure-form pure-form-aligned" do %>
```
Presumably this is what's being used on the `:new` template.  You should use a `before_action` to create a `PhotoProof.new` object, and then use the form_helper for @job, @photo_proof here too.

jayfredlund commented on app/views/jobs/photo_proofs/_upload_form.html.erb:
```diff
@@ -0,0 +1,6 @@
+<%= form_tag job_photo_proofs_path(job_id: @job.id), id: "photo-proof-upload-form", enctype: 'multipart/form-data', class: "pure-form pure-form-aligned" do %>
+  <%= file_field_tag :file, accept: "image/*", capture: "camera"  %>
+  <%= button_tag type: :submit, id: "photo-proof-submit-button", class: "pure-button button-primary" do %>
+    <i class="fa fa-camera"></i> Send Photo Proof
```
You should be able to use the `icon` helper here too:
```ERB
<%= icon('camera') %> Send Photo Proof
```

jayfredlund commented on app/views/jobs/photo_proofs/approve.html.erb:
```diff
@@ -0,0 +1,7 @@
+<div class="pure-g">
+  <div class="pure-u-1">
+    <h2>Approve Photo Proof</h2>
+    <%= render "photo_proof", photo_proof: @photo_proof %>
```
You probably don't need to use local variables here...instead just favoring the instance variable since you're not rendering this template multiple times with different objects.

jayfredlund commented on app/views/jobs/photo_proofs/index.html.erb:
```diff
@@ -0,0 +1,7 @@
+<% if @job.photo_proofs.any? %>
+<h2>Photo Proofs</h2>
+  <% @job.photo_proofs.each do |p| %>
+    <%= render "jobs/photo_proofs/photo_proof", photo_proof: p %>
```
Nevermind my last comment.  You are looping through it here :)

jayfredlund commented on app/views/jobs/photo_proofs/index.html.erb:
```diff
@@ -0,0 +1,7 @@
+<h1><%= @photo_proof_details['job_name'] %></h1>
+<% if @job.photo_proofs.any? %>
```
`@job.photo_proofs` would probably be a good candidate to be set in a `before_action` in the controller:

```Ruby
def get_photo_proofs
  @photo_proofs = @job.photo_proofs
end
```

Also, do we need to use `any?` since this is a collection of objects?  It should just be objects or nil.

jayfredlund commented on config/environments/development.rb:
```diff
@@ -42,6 +45,12 @@
+  config.kotis_uri = "https://kotisdesign.com"
+
+  # Aws defaults
+  config.aws_s3_bucket = "kotis-dev"
```
AWS settings like the Region go in the secrets file.

Carrierwave settings (bucket names, etc) go in the carrierwave config.

Though it doesn't look like we're using these anyway..

jayfredlund commented on config/initializers/aws.rb:
```diff
@@ -0,0 +1,5 @@
+Aws.config.update({
```
I don't think we need this specific config file now that we've remove the manual s3 uploader, right?

jayfredlund commented on config/initializers/carrierwave.rb:
```diff
@@ -0,0 +1,17 @@
+    aws_secret_access_key: Rails.application.secrets.s3_secret_key,
+    region:                Rails.application.config.aws_s3_region
+  }
+  config.fog_directory = Rails.application.config.aws_s3_bucket
```
I'd just copy the carrierwave config file from any of the other apps so it's the same across everything.

This would specify the production bucket here (rather than the environment config files), and would pull the region in from the secrets file as well.

We also shouldn't need to specify the `store_dir` on line 15 below.

ilynne commented:
OK, comments addressed. Note that comments related to the "Approve/Reject" function are left as-is and will be addressed in issue #3.

ilynne commented:
Removed approve/reject views, methods, routes, specs. Also removed update methods, routes, specs, since there is no reason for them without approval. These will be added back in issue #3

ilynne commented:
Comments re-scanned twice. Found a typo and some unnecessary asset files.

New screenshots:

Redirect to show page, not new, after successful upload.
![kotis_print_shop](https://cloud.githubusercontent.com/assets/3128785/9535375/842f7b22-4cd5-11e5-9bd4-5904638000cc.png)

Email with no approval link, since that will be addressed in issue #3 
![photo_proof_for_job_115413](https://cloud.githubusercontent.com/assets/3128785/9535391/a1fa6810-4cd5-11e5-9ff7-b97e77f508e9.png)

Photo Proofs index page, small screen. No "Approved"/"Rejected"/"Awaiting Approval" messages. They will be added back in issue #3 
![kotis_print_shop](https://cloud.githubusercontent.com/assets/3128785/9535407/bcddcc58-4cd5-11e5-8348-9adb13cc1bb8.png)
